### DISCUSSION TITLE
Allow specifying fall back header cache backends/compress algorithms

### DISCUSSION THREAD

**Discussion by whitney-cumber** [neomutt/neomutt#4518] [Development]
## What would you like to be able to do?

I'd like to give a "priority order" for the header cache backends.  Use case is to have the same configuration file for neomutt even when different compile-time features are present, namely header cache backends.

The current `$header_cache_backend` allows only for one backend to be specified (or to use Neomutt's builtin priority list).  If that backend is not compiled in, then the config fails.

A similar problem arises with `$header_cache_compress_method`.

## What needs to change in NeoMutt?

Two solutions pop into my mind:

1) make `$header_cache_backend`/ `$header_cache_compress_method` a list of strings instead of strings (no clue what a good separator char would be)
2) export the compiled in header cache backends / compressions as compile options

In the later case the priority list can then be achieved by
```
ifdef tdb set header_cache_backend = tdb    # fall back if everything else fails
ifdef qdbm set header_cache_backend = qdbm  # silver medal goes here
ifdef ldbm set header_cache_backend = ldbm  # top choice
```

**Comment by flatcap** [comment]
> `ifdef ldbm set header_cache_backend = ldbm  # top choice`

Hmm... I don't think that'll work.
We could add the stores' names to the existing ifdef tests:
https://github.com/neomutt/neomutt/blob/5fcc6c9e9f606f47291066ad1022fe805545640a/command_parse.c#L504-L509

My first thought was: this might be quite a bit of work.  But, if you're offering :-)

Next, I thought this ought not to be too difficult.
We turns the config options into `DT_SLIST`; we use the validator to check them, then tweak the hcache to select one.
(The config string type supports different separators)
https://neomutt.org/code/config_slist.html

But, if a store isn't compiled-in, then the name simply doesn't exist in the code.
We wouldn't be able to distinguish compiled-out from rubbish:
```
set header_cache_backend = rubbish,lmdb
```
Perhaps that doesn't matter.

Another possibility is to reorder `StoreOps[]` to put the modern databases first, e.g. lmdb, tdb, rocksdb:
https://github.com/neomutt/neomutt/blob/5fcc6c9e9f606f47291066ad1022fe805545640a/store/store.c#L50-L53

**Comment by whitney-cumber** [comment]
> My first thought was: this might be quite a bit of work. But, if you're offering :-)

Sorry but I've never learned a C style language.  So that bit of work might be quite to much for me. :-(

Shouldn't it be enough to include something like
```
ifdef HAVE_LMDB
  { "lmdb", 1 },
else
  { "lmdb", 0 },
endif
```
in version.c like here

https://github.com/neomutt/neomutt/blob/5fcc6c9e9f606f47291066ad1022fe805545640a/version.c#L160-L170

to make `ifdef lmdb` in neomuttrc work?

---

>[making $header_cache_backend a list]
>But, if a store isn't compiled-in, then the name simply doesn't exist in the code.
>We wouldn't be able to distinguish compiled-out from rubbish:
>
>set header_cache_backend = rubbish,lmdb
>
>Perhaps that doesn't matter.

So if it's not compiled in the above `set` command would error out?  Then this approach seems less useful to me than guarding the setting by `ifdef`s.  If we allow rubbish then other garbage would also be accepted which I find suboptimal as reporting unknown settings to the user is a good practice (catches lots of typos, ...).

---

> Another possibility is to reorder StoreOps[] to put the modern databases first, e.g. lmdb, tdb, rocksdb:

Saner defaults are always welcome, so this should be done (unless backwards compatibility jumps in).
However, different people might have different opinions about what database is "better", so they might prefer a slightly different order than other people, i.e. the underlying problem of being able to specify the preference isn't solved that way.

**Comment by flatcap** [comment]
> Sorry but I've never learned a C style language

Oh!  Sorry for the assumption :-)
I'm happy to mentor, should you wish to try some easy tasks.

> ifdef HAVE_LMDB
>   { "lmdb", 1 },
> else
>   { "lmdb", 0 },
> endif

We wouldn't want to add them to the 'features' list, because that would just create more clutter, but a separate store list would be ok.
We already list the stores (`neomutt -v`):

```
storage: tokyocabinet, kyotocabinet, qdbm, rocksdb, gdbm, bdb, tdb, lmdb
```

With a list of **all** the stores, we could correctly validate `set header_cache_backend = rubbish,lmdb`

> Saner defaults are always welcome,

Yeah.
Putting `lmdb` first is probably a good idea, anyway.
`tokyocabinet` is the default from `mutt` from 20 years ago -- it's not competitive with modern DBs.
https://neomutt.org/contrib/hcache-bench

> so this should be done (unless backwards compatibility jumps in).

Actually, this won't be a problem.

The header cache is a binary dump of the in-memory C `struct`s: `Email`, etc
This means that the cache is highly sensitive to code changes.
This means that most NeoMutt releases cause the invalidation of the entire cache.
So, if we changed the default store to `lmdb`, one release, many people would benefit.

> different people might have different opinions about what database is "better"

True.
Not just the users -- few of them build their own NeoMutt.
The package maintainers get to choose which DBs are actually compiled-in.
